### PR TITLE
feat: [PoC] Support disabling PHP-CS-Fixer for single lines and rules

### PR DESCRIPTION
**Pull request by windaishi**
This PR is a proof of concept on wether it is possible to disable PHP-CS-Fixer for single lines in the code and for single rules.

Please note that this PR is only a proof of concept and therefore no attention has been paid to details yet.

The idea behind this is to modify all tokes between to marker tokens, so that they won't be changed by fixer. The marker-tokens are simple comments. The modification is currently only by done setting a special ID to a token so the fixer will hopefully leave them alone. After a fixer has applied changes to the remaining tokes, the tokens will be changed back to their original state. Now the next fixer can work with the tokens again.

## Syntax:

* Disable all rules: `// phpcsfixer:disable`
* Disable some rules: `// phpcsfixer:disable rule1,rule2`
* Enable rules again: `// phpcsfixer:enable`

## Example

```php
public function containsElementEqualTo(mixed $search): bool
{
    // phpcsfixer:disable strict_comparison
    return $this->first(fn($element) => $element == $search) !== null;
    // phpcsfixer:enable
}
```

Related to: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/4512

### CODE REVIEW

**Comment by Wirone** [comment]
Thank you @windaishi for your effort around this difficult feature! Looking at the code alone I don't know if it's actually working - please add tests that show this in practice 🙂.

**Comment by windaishi** [comment]
@Wirone I added some tests.

**Comment by kubawerlos** [comment]
TBH I cannot imagine how this can happen, it would be a huge effort to adjust all the fixers. To show what I mean, please add these code samples to tests:

- config `['array_syntax' => true]`
  ```php
   <?php
   $array = array(
       1,
       // phpcsfixer:disable
       2,
   );
   // phpcsfixer:enable
   ```

- config `['ordered_imports' => true]`
  ```php
   <?php
   use D;
   // phpcsfixer:disable
   use C;
   use B;
   // phpcsfixer:enable
   use A;
   ```

- config `['statement_indentation' => true]`
  ```php
   <?php
   $goodIntentation = 7;
   // phpcsfixer:disable
                                   $toMuchIndentation = foo(
                                                   $x,
                                                   $y,
                                                   // phpcsfixer:enable
                                                   $z,
                                   );
   ```

**Comment by windaishi** [comment]
What would he the expected behavior for the test cases you mentioned? Should they fail the fixing process?

**Comment by kubawerlos** [comment]
That's a very good question.

> Should they fail the fixing process?

Currently, the sole idea of PHP CS Fixer is to always be able to fix the file, and only to fail on `--dry-run` check.

**Comment by windaishi** [comment]
The fixing process fails for example when a file has a syntax error. (As far as I know.)
The example
```
<?php
$array = array(
   1,
   // phpcsfixer:disable
   2,
);
// phpcsfixer:enable
```
Could be interpreted as one.

**Comment by kubawerlos** [comment]
> when a file has a syntax error

Then I'd say the fixing process doesn't start. By "fixing process" I mean the process of modifying the collection of tokens created by `PhpCsFixer\Tokenizer\Tokens`, which never fails now.

**Comment by windaishi** [comment]
I added 3 more test for the cases you mentioned. I guess the fixer is doing pretty good for this special cases. Only the array may be a problem.

I'd then suggest to tokenizes the enable/disable comments as their own token and check that enable and disable comments are on the same code block level. If not, the comments are just treated as normal comments.

File: src/AbstractFixer.php [hunk 1]
```diff
@@ -71,9 +72,50 @@ final public function fix(\SplFileInfo $file, Tokens $tokens): void
...
+            }
+
+            if ($isIgnored && $token->isGivenKind(T_COMMENT) && '// phpcsfixer:enable' === $token->getContent()) {
+                $isIgnored = false;
+
+                continue;
+            }
+
+            if ($isIgnored) {
+                $dummyToken = new Token([999999, $token->getContent()]);
```
Comment (keradus) [inline src/AbstractFixer.php:93]: this is extremely dangerous. here, the invalid token is created. It is affecting traversing logic of each rules, as they do things like `->getNextTokenOfKind(KIND_ID)`, and may crash if not found.

I'm unable to, and see it as huge effort, to ensure each rule can handle invalid token in any part of codebase and still work properly with it.

Comment (keradus) [reply src/AbstractFixer.php:93]: example 1:

```php
<?php

array(
    1   ,
    2   ,
// phpcsfixer:disable
);
// phpcsfixer:enable
```

```
ker@d ~/github/FriendsOfPHP/PHP-CS-Fixer λ ./php-cs-fixer check --diff x.php --rules no_whitespace_before_comma_in_array -vvv
PHP CS Fixer 3.41.2-DEV Long Night by Fabien Potencier and Dariusz Ruminski.
PHP runtime: 8.3.0
Loaded config default from "/Users/d.ruminski/github/FriendsOfPHP/PHP-CS-Fixer/.php-cs-fixer.dist.php".
Paths from configuration file have been overridden by paths provided as command arguments.
E                                                                                                                                1 / 1 (100%)
Legend: .-no changes, F-fixed, S-skipped (cached or empty file), I-invalid file syntax (file ignored), E-error

Found 0 of 1 files that can be fixed in 0.003 seconds, 14.000 MB memory used

Files that were not fixed due to errors reported during fixing:
   1) /Users/d.ruminski/github/FriendsOfPHP/PHP-CS-Fixer/x.php

                                    
        [UnexpectedValueException]  
        Missing block "end".        
                                    

      PhpCsFixer\Tokenizer\Tokens->findOppositeBlockEdge()
        in /Users/d.ruminski/github/FriendsOfPHP/PHP-CS-Fixer/src/Tokenizer/Tokens.php at line 445
      PhpCsFixer\Tokenizer\Tokens->findBlockEnd()
        in /Users/d.ruminski/github/FriendsOfPHP/PHP-CS-Fixer/src/Fixer/ArrayNotation/NoWhitespaceBeforeCommaInArrayFixer.php at line 89
      PhpCsFixer\Fixer\ArrayNotation\NoWhitespaceBeforeCommaInArrayFixer->fixSpacing()
        in /Users/d.ruminski/github/FriendsOfPHP/PHP-CS-Fixer/src/Fixer/ArrayNotation/NoWhitespaceBeforeCommaInArrayFixer.php at line 64
      PhpCsFixer\Fixer\ArrayNotation\NoWhitespaceBeforeCommaInArrayFixer->applyFix()
        in /Users/d.ruminski/github/FriendsOfPHP/PHP-CS-Fixer/src/AbstractFixer.php at line 104
      PhpCsFixer\AbstractFixer->fix()
        in /Users/d.ruminski/github/FriendsOfPHP/PHP-CS-Fixer/src/Runner/Runner.php at line 173
      PhpCsFixer\Runner\Runner->fixFile()
        in /Users/d.ruminski/github/FriendsOfPHP/PHP-CS-Fixer/src/Runner/Runner.php at line 114
      PhpCsFixer\Runner\Runner->fix()
        in /Users/d.ruminski/github/FriendsOfPHP/PHP-CS-Fixer/src/Console/Command/FixCommand.php at line 302
      PhpCsFixer\Console\Command\FixCommand->execute()
        in /Users/d.ruminski/github/FriendsOfPHP/PHP-CS-Fixer/vendor/symfony/console/Command/Command.php at line 326
      [ ... ]

```

Comment (keradus) [reply src/AbstractFixer.php:93]: example 2:
```php
<?php

array(
    1,
    2,
    3,
    // phpcsfixer:disable
    array(
        4,
        5
    // phpcsfixer:enable
    ),
    6
);
```
```
ker@d ~/github/FriendsOfPHP/PHP-CS-Fixer λ ./php-cs-fixer check --diff x.php --rules array_syntax -vvv
PHP CS Fixer 3.41.2-DEV Long Night by Fabien Potencier and Dariusz Ruminski.
PHP runtime: 8.3.0
Loaded config default from "/Users/d.ruminski/github/FriendsOfPHP/PHP-CS-Fixer/.php-cs-fixer.dist.php".
Paths from configuration file have been overridden by paths provided as command arguments.
E                                                                                                                                1 / 1 (100%)
Legend: .-no changes, F-fixed, S-skipped (cached or empty file), I-invalid file syntax (file ignored), E-error

Found 0 of 1 files that can be fixed in 0.005 seconds, 14.000 MB memory used

Files that were not fixed due to errors reported during linting after fixing:
   1) /Users/d.ruminski/github/FriendsOfPHP/PHP-CS-Fixer/x.php

                                                                            
        [PhpCsFixer\Linter\LintingException]                                
        Parse error: Unclosed '(' on line 8 does not match ']' on line 12.  
                                                                            

      Applied fixers: array_syntax
      ---------- begin diff ----------
--- /Users/d.ruminski/github/FriendsOfPHP/PHP-CS-Fixer/x.php
+++ /Users/d.ruminski/github/FriendsOfPHP/PHP-CS-Fixer/x.php
@@ -1,6 +1,6 @@
 <?php
 
-array(
+[
     1,
     2,
     3,
@@ -9,6 +9,6 @@
         4,
         5
     // phpcsfixer:enable
-    ),
+    ],
     6
 );
\ No newline at end of file

      ----------- end diff -----------
```

**Comment by windaishi** [comment]
I agree that the examples you provided are all problematic somehow. Anyway I'd like the feature to happen. To continue my work, I need to know what the expected behavior should be.

I use the following code example to discuss possible behaviors.
```php
// phpcsfixer:disable
array(
// phpcsfixer:enable
    4,
);
```

This example should be seen as an invalid usage of the `// phpcsfixer:disable` directive. To handle that we have several possibilities:

1. Fix the code as usual, but keep the code part, where the fixer is disabled, as is
    ```php
    // phpcsfixer:disable
    [
    // phpcsfixer:enable
        4,
    );
    ```
    
    Problem: This leads to syntax errors in files.

2. Fail the fixing process for the file with an error. Inform the user that the "syntax" of the file is invalid and therefore it cannot be fixed. 
3. Ignore the `// phpcsfixer:disable` and `// phpcsfixer:enable` directives.
    ```php
    // phpcsfixer:disable
    array(
    // phpcsfixer:enable
        4,
    );
    ```

I'd personally prefer 2. because it is a good compromise between 1. and 3.

**Comment by keradus** [comment]
Thank you for the contribution. I do see this topic very valid. However, after thorough review, it seems that the current implementation approach poses significant risks to our internal tokenization system and project safeguards. Specifically, the creation of invalid tokens put at risk tokens traversing, finding brackets siblings etc and linting issues.

We have strict guidelines in place to ensure the stability and security of our project, and unfortunately, this approach doesn’t align with those standards.

I would suggest exploring alternative approaches that wouldn’t compromise the internal tokenization system. If you have any questions or need further clarification, let’s discuss this to find a more suitable solution together.
Idea to kick-start with:
- have a file-level comment that disable rule per whole file (we already agreed to be good to go with)
- have a counter of violations / baseline, to allow only some of in-file violations, without explicitly tagging how many (yet, useful only for `check` command, probably `fix` command will still need to skip whole file)

**Comment by Wirone** [comment]
> have a file-level comment that disable rule per whole file (we already agreed to be good to go with)

I believe it should be done via config, like in ECS (see #3871). This is an explicit information that runner should have _before_ specific file is passed to the fixer(s), instead of deciding about it after parsing file's comments.

**Comment by windaishi** [comment]
@keradus Thank you for your feedback. I will abandon my approach then. I will continue my work with one of your suggestions. I will create an issue for that if none exists yet, if that is okay for you.

@Wirone I don't really like keeping such ignore-rules in the configuration. Why a rule is ignored is strongly dependent on the file contents. If you for example rename a file, the rule should still be ignored because the content is the same. Keeping a list of ignored files in the config imho is hard to maintain.

**Comment by windaishi** [comment]
If you want you can assign me to this ticket: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/3871

**Comment by Wirone** [comment]
@windaishi IMHO it's not hard to maintain because if you change the file path it suddenly goes out of the ignore rules and you'll spot it in the tool's output 🙂. Also, with renaming in IDE the path can be automatically updated. The main advantage of having this in config is performance - you have explicit rules and don't need to process the file just to determine it should be ignored.

**Comment by keradus** [comment]
from UX perspective, I'm more to have "ignore rule X in file Y" in file itself rather than centralised in project configuration. Could be that different ppl have different preference here and maybe over time we should have both. I would prefer to start with in-file one.

If we implement skip-of-rule-per-file in Runner, I would see no performance difference. The fixer (rule) class won't fire with it's logic for both approaches (project config / in-file config)

**Comment by Wirone** [comment]
The logic won't run, but you still need to parse each file just to check whether it should be processed. For me it does not make sense, but whatever..

**Comment by keradus** [comment]
for whole tool, we parse file once (plus linting), regardless if it's single rule to process the file, or 300 rules ;) as long as at least single rule is executed, it shall not affect the performance.

**Comment by Wirone** [comment]
And you could skip that parse and linting, but you want to add logic for determining ignore rules instead 🤷‍♂️. It's 3 operations difference for each file.

**Comment by keradus** [comment]
as long as at least one rule will process the file, we cannot skip parse/lint.
in any variant, we need to grab exception list from somewhere - i find this task to be same complexity regardless what is source of the exception list - config, baseline or file itself.

**Comment by Wirone** [comment]
> as long as at least one rule will process the file, we cannot skip parse/lint.

@keradus Your approach for each file during processing:

- lint the file to determine if the syntax is correct
- parse the file to get exclude list
- run each configured rule on the file _unless_ it's marked as ignored (it would come from the previous step)
- lint the file after applying fixes to determine we did not break the syntax

My approach:

- lint the file to determine if the syntax is correct
- run each configured rule on the file _unless_ it's marked as ignored (it would come from the resolved config)
- lint the file after applying fixes to determine we did not break the syntax

So yeah, linting must be done anyway, but we could get rid of file parsing to get ignore rules, which means as many I/O operations less as files configured for processing. Besides that, I see ignoring rules as some kind of "infrastructure" code, that should not live in the source code files.

**Comment by keradus** [comment]
`parse the file to get exclude list`

if file will be processed by at least single rule, we need to parse the file.
And then effort to get special file-level comment for exceptions, or config file glob for exception, is similar to execute.

> Besides that, I see ignoring rules as some kind of "infrastructure" code, that should not live in the source code files.

I see it close to "code style" of the code which is closest to the source code file itself rather rather than glob path.
But you know, I believe that strongly depends of use case, where one will want to disable whole PHP 8.2 migration for subpackage that has to be PHP8 compatible, and other one will like to disable protected_to_private for single class due to special logic within that file.

Thus... [ref](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7569#issuecomment-1868511940)
> from UX perspective, I'm more to have "ignore rule X in file Y" in file itself rather than centralised in project configuration. Could be that different ppl have different preference here and maybe over time we should have both. I would prefer to start with in-file one.

**Comment by Wirone** [comment]
It's still logic that needs to be created, maintained and of course executed before applying the rules. I find it superfluous, because you could easily get map of ignore rules from the config and just check it before applying each rule.

The fact that you could ignore the whole directories (at any nesting level) or subsets of files matching the pattern makes it a better option, because with one single change you can fix the problem you encounter e.g. after upgrading to a newer version. With your approach it would require modifications in multiple (sometimes dozens of) files, which I find totally pointless. Maintaining this would be a pain in the ass 😉.

**Comment by keradus** [comment]
@Wirone , I'm failing to see the point of this discussion. What is the goal of it?

I shared 2 different use case, in which 2 different variants can work better. Different ppl also shared different preference. Msg from days ago suggest to have both variants available.
I will compare it to rule options. We have option that allow users to go with their preference, because sometime "one for all" won't work. Not sure if we can come with any other outcome here.

**Comment by Wirone** [comment]
The point is to discuss pros and cons. In my opinion your arguments were not good enough, just like @windaishi's (addressed [here](#issuecomment-1867662967)). Personally I don't see a point in defining file-wide ignore rules on file level because of reasons provided above, but whatever.

**Comment by keradus** [comment]
so that's exactly the point - it's OK to have ppl with different preferred way of going.
we have options for everything, rules, reports, progress indicator, why not this one? ;)

**Pull request closed without merging**